compiler/java/src/main/java/com/c/lexer/Lexer.java

```diff
@@ -1,29 +1,635 @@
+            throw new InvalidCommentsIdentifierException("You provided an identifier to close block comments but not one to open them");
+        if (commentLineIdentifier.isEmpty() && commentBlockOpenIdentifier.isEmpty())
+            throw new InvalidCommentsIdentifierException("You must provide at least line comments or block comments identifier");
+
+        // Construct conditional comments identifiers based on comment identifiers of the underlying language (e.g. Java)
+        this.commentLineIdentifier = commentLineIdentifier.isEmpty() ? "" : commentLineIdentifier + "?";
+        this.commentBlockOpenIdentifier = commentBlockOpenIdentifier.isEmpty() ? "" : commentBlockOpenIdentifier + "?";
+        this.commentBlockCloseIdentifier = commentBlockCloseIdentifier;
+        this.commentPayloadIdentifier = commentLineIdentifier; // might be "", that's ok and considered in the grammar
+
+        // Calculate max look ahead
+        List<Integer> commentIdentifierLengths = new ArrayList<>();
+        commentIdentifierLengths.add(commentLineIdentifier.length());
+        commentIdentifierLengths.add(commentBlockOpenIdentifier.length());
+        commentIdentifierLengths.add(commentBlockCloseIdentifier.length());
+        commentIdentifierLengths.add(commentPayloadIdentifier.length());
+        // We need one character more than the max for the method isLookAheadCommentBlockCloseIdentifierWithBrace()
+        int maxLookAhead = Collections.max(commentIdentifierLengths) + 1;
+
+        // Construct FileReader
+        this.reader = new FileReader(file, maxLookAhead);
+
+        // Set context
+        currentContext = isSingleStatement || isLookAheadCommentLineIdentifier() || isLookAheadCommentBlockOpenIdentifier()
+                ? Context.SECTION : Context.ARBITRARY;
+
+        advance();
     }
 
-    public void advance() {
-        // TODO
+
+    // ----------------------------------------- Look ahead ------------------------------------------------------------
+
+    /**
+     * Looks ahead one token without consuming it.
+     *
+     * @return the next token
+     */
+    public Token lookAhead() {
+        return nextToken;
     }
 
-    // TODO
+
+    // --------------------------------------- Expect ------------------------------------------------------------------
+
+    /**
+     * Checks if the next Token is the expected Token. Advances one token after a successful check.
+     *
+     * @param t expected Token (to compare the next Token against)
+     * @throws UnexpectedTokenException if the next Token ís not equal to the expected Token
+     * @throws UnexpectedCharException  if a char - read by the FileReader - was not the expected one
+     * @throws UnknownCharException     if a char cannot be processed to a valid Token
+     */
+    public void expect(Token t) throws UnexpectedTokenException, UnexpectedCharException, UnknownCharException {
+        if (!nextToken.equals(t))
```

Comment:
That's funny, I wrote the opposite for your implementation ;) That's because I haven't implemented the Parser yet. Once I'll get there, I will add another `expect()` method to expect a token type without a specific position. 👍

Reply:
Jep 😆. In this case this is completely fine.